maestro-ios-xctest-runner/maestro-driver-iosUITests/Routes/Handlers/ViewHierarchyHandler.swift

```diff
@@ -0,0 +1,182 @@
+        category: String(describing: Self.self)
+    )
+
+    func handleRequest(_ request: FlyingFox.HTTPRequest) async throws -> HTTPResponse {
+        guard let requestBody = try? JSONDecoder().decode(ViewHierarchyRequest.self, from: request.body) else {
+            let errorInfo = [
+                "errorMessage": "incorrect request body provided",
+            ]
+            let body = try? JSONEncoder().encode(errorInfo)
+            return HTTPResponse(statusCode: .badRequest, body: body ?? Data())
+        }
+
+        let appId = requestBody.appId
+
+        do {
+            let viewHierarchy = try logger.measure(message: "View hierarchy snapshot for \(appId)") {
+                try getViewHierarchy(appId: appId)
+            }
+
+            let body = try JSONEncoder().encode(viewHierarchy)
+
+            // TODO: Remove debug print here
+            print(String(data: body, encoding: .utf8)!)
+
+            return HTTPResponse(statusCode: .ok, body: body)
+        } catch {
+            print(error)
+            let errorInfo = [
+                "errorMessage": "Snapshot failure while getting view hierarchy. Error: \(error)",
+            ]
+            let body = try? JSONEncoder().encode(errorInfo)
+            return HTTPResponse(statusCode: .internalServerError, body: body ?? Data())
+        }
+    }
+
+    func getViewHierarchy(appId: String) throws -> AXElement {
+        SystemPermissionHelper.handleSystemPermissionAlertIfNeeded(springboardApplication: springboardApplication)
+
+        // Fetch the view hierarchy of the springboard application
+        // to make it possible to interact with the home screen.
+        // Ignore any errors on fetching the springboard hierarchy.
+        let springboardHierarchy: AXElement?
+        do {
+            springboardHierarchy = try elementHierarchy(xcuiElement: springboardApplication)
+        } catch {
+            print("Springboard hierarchy failed to fetch: \(error)")
+            springboardHierarchy = nil
+        }
+
+        let appHierarchy = try appHierarchy(XCUIApplication(bundleIdentifier: appId))
+
+        return AXElement(children: [
+            springboardHierarchy,
+            appHierarchy,
+        ].compactMap { $0 })
+    }
+
+    func appHierarchy(_ xcuiApplication: XCUIApplication) throws -> AXElement {
+        AXClientSwizzler.overwriteDefaultParameters["maxDepth"] = maxDepth
+        return try elementHierarchyWithFallback(element: xcuiApplication)
+    }
+
+    func elementHierarchyWithFallback(element: XCUIElement) throws -> AXElement {
+        do {
+            var hierarchy = try elementHierarchy(xcuiElement: element)
+            let hierarchyDepth = hierarchy.depth()
+
+            if hierarchyDepth < maxDepth {
+                logger.info("Hierarchy dept below maxdepth \(hierarchyDepth)")
+                return hierarchy
+            }
+
+            // When the hierarchy depth is equal to maxDepth, it is unknown if
```

Comment:
did you manage to reproduce this case? my expectation was that `xcuiElement.snapshot().dictionaryRepresentation` always gives us complete hierarchy and crashes with `kAXErrorIllegalArgument` exception when its depth limits are reached

also it should be `is greater or equal to maxDepth` here looking at the code

Reply:
This code has been removed since we don't need to set maxDepth